### DISCUSSION TITLE
Add support for more NumPy routines

### DISCUSSION THREAD

**Discussion by EmilyBourne** [pyccel/pyccel#1534] [Ideas]
This discussion contains requested NumPy routines so we can gauge interest and consider any potential complexities:

- [ ] np.genfromtxt
- [ ] np.argmax
- [ ] np.trapz
- [ ] np.column_stack
- [ ] np.savetxt
- [ ] np.true_divide
- [ ] np.outer
- [ ] np.newaxis
- [ ] np.reshape
- [ ] np.cross

**Comment by VaradRengaraj** [comment]
Hi @EmilyBourne may be i can work on np.outer & np.argmax.

**Reply by EmilyBourne** [reply]
Feel free, but as a word of warning it is probably better to avoid writing too much C code as we are currently working towards changing the implementation of ndarrays in C

**Reply by VaradRengaraj** [reply]
@EmilyBourne Can you tell me , how does this new ndarrays implementation look like if it is not in C?

**Reply by EmilyBourne** [reply]
It will still be in C. We plan to use https://github.com/Stclib/STC/
It's just that changing the implementation will mean updating all code written in C so we want to avoid increasing this codebase before that change. STC also implements some algorithms so we may be able to use those for some implementations

**Comment by vadter** [comment]
Are there plans to support numpy structured arrays?

**Reply by EmilyBourne** [reply]
Structured arrays like this: https://numpy.org/doc/stable/user/basics.rec.html ?

I think this was requested once a few years ago. We started looking into it but from what I remember developments trailed off because no-one on our end uses structured arrays in their work and the requester stopped replying. We never got a test case to base a PoC off.

As we release v2.0 on Saturday this is definitely the right time to make requests though if this is something you would be interested in.

**Reply by EmilyBourne** [reply]
@vadter 
If you are interested in this we would need an example Python code illustrating a basic use case.

If you have any ideas about what the equivalent code should be in C and/or Fortran then this would also be appreciated.

**Reply by vadter** [reply]
Emily,
Using structured arrays will only make things easier, as I understand it, for example, when passing arguments to functions. But this will have little effect on the efficiency of the algorithms. Therefore, I think this task is not as high a priority as the others.
I will try to post an example code and try to explain my position as soon as I have the opportunity.

**Reply by vadter** [reply]
```
import numpy as np

def f(M) -> 'float[:]':
    ret = np.empty(M[0]['I'].size, dtype = np.complex128)

    for i in range(ret.size):
        ind = M[0]['I'][i]
        ret[i] = M[0]['A'][ind] * M[0]['R'][ind]

    return ret


# Define a structured data type with a fixed-size array field
dt = np.dtype([
    ('R', (float, 5)), # A subarray of 5 floats
    ('A', (complex, 5)), # A subarray of 5 complexes
    ('I', (int, 3))     # A subarray of 5 integers
])

# Create an empty structured array with 1 element
M = np.zeros(1, dtype = dt)

# Assign values
M[0]['R'] = np.array([1.1, 2.3, 4.5, 2.6, 9.1], dtype = np.float64) # coords
M[0]['A'] = np.array([10.1, 20.3, 40.5, 20.6, 90.1], dtype = np.complex128) # data
M[0]['I'] = np.array([1, 3, 4], dtype = np.int64) # indexes


r = f(M)

print(r)
```

**Reply by EmilyBourne** [reply]
@vadter Generally arrays of structs are worse for performance than structs of arrays. Given that Pyccel handles classes is there an argument to use structured arrays here rather than:
```python
import numpy as np

class MType:
    def __init__(self, n : int):
        self.R = np.empty((n,5), dtype = float)
        self.A = np.empty((n,5), dtype = complex)
        self.I = np.empty((n,3), dtype = int)

def f(M : MType) -> 'float[:]':
    ret = np.empty(M.I.shape[1], dtype = np.complex128)

    for i,ind in enumerate(M.I[0,:]):
        ret[i] = M.A[0,ind] * M.R[0,ind]

    return ret


# Create an empty structured array with 1 element
M = MType(1)

# Assign values
M.R[0,:] = np.array([1.1, 2.3, 4.5, 2.6, 9.1], dtype = np.float64) # coords
M.A[0,:] = np.array([10.1, 20.3, 40.5, 20.6, 90.1], dtype = np.complex128) # data
M.I[0,:] = np.array([1, 3, 4], dtype = np.int64) # indexes


r = f(M)

print(r)
```
?

**Reply by vadter** [reply]
Emily, thank you, this is what I want )
